### DISCUSSION TITLE
Error: Unsupported Algorithm When Extracting Public Certificate from PKCS#12 File

### DISCUSSION THREAD

**Discussion by EthanTang-Zq** [openssl/openssl#23089] [Q&A] [answered]
**Problem Description:**
When running the following command:

`openssl pkcs12 -in testit.aggsimida.com-Entire_certificate-030332.p12 -clcerts -nokeys -out public.crt `
I encountered the following error:

```
Enter Import Password:
Error outputting keys and certificates
801B5126747F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:../crypto/evp/evp_fetch.c:349:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()
```

**Version Information:**
OpenSSL Version: OpenSSL 3.0.2 15 Mar 2022 (Library: OpenSSL 3.0.2 15 Mar 2022)
Operating System: Ubuntu 22.04

**Steps to Reproduce:**
Run the above OpenSSL command.
Enter the import password when prompted.

**Expected Behavior:**
Expecting to successfully extract the public certificate without encountering errors.

**Actual Behavior:**
Encountered an error related to the unsupported algorithm.

**Comment by nhorman** [comment]
Can you post the output of the following command please:
```
./configdata.pm --dump
```
I suspect that your openssl build doesn't have the rc2 algorithm included in your configuration

**Comment by EthanTang-Zq** [comment]
Please run this command in which directory. I am using openssl in the Ubuntu software package.

**Comment by mattcaswell** [comment, marked as answer]
> 801B5126747F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:../crypto/evp/evp_fetch.c:349:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()

This tells you that the file has been encrypted with an unsupported encryption algorithm (RC2-40-CBC). That algorithm is now considered legacy and insecure. The algorithm is still available in our legacy provider. Please try adding "-legacy" to your command line.

**Reply by boniface1994** [reply]
thanks you very match

**Reply by HodanekHonza** [reply]
OMG THANKS!

**Reply by jrdaher** [reply]
big thanks

**Reply by Pankaj7988** [reply]
Thank you

**Reply by marcosfaneli** [reply]
thank you

**Reply by ghost** [reply]
Thank you very much

**Reply by MrPalloncini** [reply]
YOU WILL NEVER BE A REAL WOMEN

**Reply by jonathandacruz** [reply]
Thaanks!

**Reply by matthaighdj** [reply]
This worked for me too!

**Reply by anhtukhtn** [reply]
thank you. worked for me

**Reply by jerrickhakim** [reply]
@mattcaswell I can't tell you how thankful I am for your answer! Much love ✌️.

**Reply by simionattovini-pro** [reply]
Some heroes don't wear capes!

**Reply by meorramzi** [reply]
Thank you very much ! worked for me !!

**Reply by SugarCRMSupport** [reply]
MANY THANKS! "-legacy" added to the CLI did the trick for me. I've not seen this before, so your help is GREATLY appreciated!

**Comment by saviour123** [comment]
With -legacy flag. This worked for me as mentioned.

`$ openssl pkcs12 -in fileWith.p12 -out Output.pem -nodes -nocerts -legacy`

**Comment by raryum** [comment]
this method didn't work for me, this time it gave another error 
openssl pkcs12 -in PfxDosyam.pfx -out certificate.crt -nodes -nocerts -legacy  

pkcs12: unable to load provider legacy
Hint: use -provider-path option or OPENSSL_MODULES environment variable.
1C230000:error:12800067:DSO support routines:win32_load:could not load the shared library:crypto\dso\dso_win32.c:108:filename(C:\Program Files\OpenSSL\lib\ossl-modules\legacy.dll)
1C230000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto\dso\dso_lib.c:147:
1C230000:error:07880025:common libcrypto routines:provider_init:reason(37):crypto\provider_core.c:950:name=legacy

When I do not type **-legacy,** I get the following error
![Screenshot_11](https://github.com/user-attachments/assets/2d2e0e2f-9c5f-40f9-ac8b-880f62de672f)

Error outputting keys and certificates
3C3C0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto\evp\evp_fetch.c:355:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()

**Reply by rutuja-dhore** [reply]
i have the same issue. have you got an update on this

**Reply by saviour123** [reply]
@raryum @rutuja-dhore I would recommend, you switch to Linux Subsystem for Windows and do it on it, My solution was for LINUX/UNIX

**Reply by He-Man321** [reply]
Hi,
I got a very similar issue when I tired this. My first question is why is Let's Encrypt\Crypt LE outputting .pfx files with an old cypher (RC2-40-CBC) and is that a problem? Using -legacy does solve the issue (see below) but are we papering over a nasty crack here and should we be finding out how to not use RC2-40-CBC in the first place?

but to resolve the issue, I found the legacy.dll in the "openssl-3.4.1\x64\lib\ossl-modules" folder and copied it in to the openssl folder and then used the -provider-path thus:
`openssl/openssl pkcs12 -in example.com.pfx -nocerts -out example.com-key.pem -nodes -password pass:password -legacy -provider-path ./openssl`
Note: because I am writing a script to do this on bulk I am calling the command from the folder above where the openssl.exe sits.

**Comment by SamTopspin** [comment]
Reinstalled/upgraded to the latest OpenSSL few days ago (coincidently) and then today encountered this error.
For me specifying the path worked like a charm, here is the command that worked.

`openssl pkcs12 -in cert1.pfx -out cert1.pem -nocerts -provider-path "C:\Program Files\OpenSSL-Win64\bin" -legacy`
